File: sort/InsertSort.cc
#include <iostream>
#include <limits>

#include "utils.h"

using namespace std;

void InsertSort(IntArray &data)
{
    IntArray temp;
    for (ArrayIter input_iter = data.begin(); input_iter != data.end(); input_iter++)
    {
        if (temp.empty() || *input_iter > temp.back())
        {
            temp.push_back(*input_iter);
        }
        else
        {
            temp.push_back(INT_MIN);
            ArrayIter output_iter;
            // 终止条件比较复杂
            for (output_iter = temp.end() - 1; *(output_iter - 1) > *input_iter && output_iter > temp.begin(); output_iter--)
            {
                *(output_iter) = *(output_iter-1);
            }

            *output_iter = *input_iter;
        }
    }
    data = temp;
}

int main(int argc, char const *argv[])
{
    IntArray data;
    GetRandomArray(data, 0, 100, 10);
    cout << "Raw input:" << endl;
    PrintArray(data);

    InsertSort(data);

    cout << "Result output:" << endl;
    PrintArray(data);

    return 0;
}

File: sort/InsertSort.py
import utils

def insert_sort(nums):
    temp = []
    for num in nums:
        if len(temp) == 0 or num > temp[-1]:
            temp.append(num)
        else:
            temp.append(num)
            i = len(temp) - 2
            while temp[i] > num and i > -1:
                temp[i+1] = temp[i]
                i -= 1
            temp[i+1] = num

    for i in range(len(nums)):
        nums[i] = temp[i]

def main():
    nums = utils.get_random(0,100,10)
    print("Raw input:")
    nums= [2,0,2,1,1,0]
    utils.print_nums(nums)
    insert_sort(nums)
    print("Result output:")
    utils.print_nums(nums)

if __name__ == "__main__":
    main()